File: job_copilot/app/core/dedup.py
"""Deduplicate job postings across sources via fuzzy matching."""
from __future__ import annotations

import re

from rapidfuzz import fuzz

from app.core.logging import get_logger
from app.models.schemas import JobPosting

log = get_logger(__name__)

SIMILARITY_THRESHOLD = 85


def normalize_company(name: str) -> str:
    name = name.lower().strip()
    name = re.sub(r"\s*(inc\.?|llc|ltd\.?|corp\.?|co\.?|,?\s*inc)\s*$", "", name)
    name = re.sub(r"[^a-z0-9\s]", "", name)
    return name.strip()


def normalize_title(title: str) -> str:
    title = title.lower().strip()
    title = re.sub(r"\s*[\(\[].*?[\)\]]", "", title)
    title = re.sub(r"\s*[-–—].*$", "", title)
    title = re.sub(r"\s+", " ", title)
    return title.strip()


def are_duplicates(a: JobPosting, b: JobPosting) -> bool:
    company_sim = fuzz.ratio(normalize_company(a.company), normalize_company(b.company))
    if company_sim < 70:
        return False
    title_sim = fuzz.ratio(normalize_title(a.title), normalize_title(b.title))
    return title_sim >= SIMILARITY_THRESHOLD


def deduplicate(jobs: list[JobPosting]) -> list[JobPosting]:
    """Remove duplicate jobs, keeping the first occurrence (by source priority)."""
    if not jobs:
        return []

    unique: list[JobPosting] = []
    for job in jobs:
        is_dup = False
        for existing in unique:
            if are_duplicates(job, existing):
                is_dup = True
                break
        if not is_dup:
            unique.append(job)

    removed = len(jobs) - len(unique)
    if removed > 0:
        log.info("dedup_complete", original=len(jobs), unique=len(unique), removed=removed)
    return unique

File: web/src/lib/dedup.ts
import type { JobPosting } from "./types";

function normalizeCompany(name: string): string {
  return name
    .toLowerCase()
    .trim()
    .replace(/\s*(inc\.?|llc|ltd\.?|corp\.?|co\.?|,?\s*inc)\s*$/i, "")
    .replace(/[^a-z0-9\s]/g, "")
    .trim();
}

function normalizeTitle(title: string): string {
  return title
    .toLowerCase()
    .trim()
    .replace(/\s*[\(\[].*?[\)\]]/, "")
    .replace(/\s*[-–—].*$/, "")
    .replace(/\s+/g, " ")
    .trim();
}

function similarity(a: string, b: string): number {
  if (a === b) return 100;
  if (!a || !b) return 0;

  const longer = a.length > b.length ? a : b;
  const shorter = a.length > b.length ? b : a;

  if (longer.length === 0) return 100;

  const costs: number[] = [];
  for (let i = 0; i <= longer.length; i++) {
    let lastValue = i;
    for (let j = 0; j <= shorter.length; j++) {
      if (i === 0) {
        costs[j] = j;
      } else if (j > 0) {
        let newValue = costs[j - 1];
        if (longer[i - 1] !== shorter[j - 1]) {
          newValue = Math.min(Math.min(newValue, lastValue), costs[j]) + 1;
        }
        costs[j - 1] = lastValue;
        lastValue = newValue;
      }
    }
    if (i > 0) costs[shorter.length] = lastValue;
  }

  return ((longer.length - costs[shorter.length]) / longer.length) * 100;
}

export function areDuplicates(a: JobPosting, b: JobPosting): boolean {
  const companySim = similarity(normalizeCompany(a.company), normalizeCompany(b.company));
  if (companySim < 70) return false;
  const titleSim = similarity(normalizeTitle(a.title), normalizeTitle(b.title));
  return titleSim >= 85;
}

export function deduplicate(jobs: JobPosting[]): JobPosting[] {
  if (!jobs.length) return [];
  const unique: JobPosting[] = [];
  for (const job of jobs) {
    if (!unique.some((existing) => areDuplicates(job, existing))) {
      unique.push(job);
    }
  }
  return unique;
}
